services/state/state.go

```diff
@@ -72,20 +87,91 @@ func (s *State) Execute(ctx types.BlockContext, tx models.Transaction) (*gethTyp
 	case models.TransactionCall:
 		res, err = bv.RunTransaction(t.Transaction)
 	default:
-		return nil, fmt.Errorf("invalid transaction type")
+		return fmt.Errorf("invalid transaction type")
 	}
 
 	if err != nil {
 		// todo is this ok, the service would restart and retry?
-		return nil, err
+		return err
 	}
 
 	// we should never produce invalid transaction, since if the transaction was emitted from the evm core
 	// it must have either been successful or failed, invalid transactions are not emitted
 	if res.Invalid() {
-		return nil, fmt.Errorf("invalid transaction %s: %w", tx.Hash(), res.ValidationError)
+		return fmt.Errorf("invalid transaction %s: %w", tx.Hash(), res.ValidationError)
 	}
 
+	if ok, errs := models.EqualReceipts(res.Receipt(), receipt); !ok {
```

Comment:
I believe that we may be able to get away without this `models.EqualReceipts` method, if we used the blocks'
- `TransactionsHashRoot`
- `ReceiptsRoot`
fields.
We do something similar:
```go
txHashes := evmTypes.TransactionHashes{}
for _, tx := range e.transactions {
	txHashes = append(txHashes, tx.Hash())
}
if e.block.TransactionHashRoot != txHashes.RootHash() {
	return nil, fmt.Errorf(
		"block %d references missing transaction/s",
		e.block.Height,
	)
}
```
And
```go
b.ReceiptRoot = gethTypes.DeriveSha(receipts, gethTrie.NewStackTrie(nil))
```

Reply:
It seems to me that it is more bullet-proof, to compare them this way. But can be done in a follow-up PR.